src/hrana.ts

```diff
@@ -0,0 +1,203 @@
+import * as hrana from "@libsql/hrana-client";
+
+import type { Config, Client, Transaction, ResultSet, InStatement } from "./api.js";
+import { LibsqlError } from "./api.js";
+import { expandConfig } from "./config.js";
+
+export * from "./api.js";
+
+export function createClient(config: Config): HranaClient {
+    const expandedConfig = expandConfig(config);
+    const url = expandedConfig.url;
+    if (url.protocol === "libsql:") {
+        url.protocol = "ws:";
+    } else if (url.protocol === "libsqls:") {
+        url.protocol = "wss:";
+    }
+
+    const client = hrana.open(url, expandedConfig.authToken);
+    return new HranaClient(client, expandedConfig.transactions);
+}
+
+export class HranaClient implements Client {
+    client: hrana.Client;
+    #transactions: boolean;
+
+    /** @private */
+    constructor(client: hrana.Client, transactions: boolean) {
+        this.client = client;
+        this.#transactions = transactions;
+    }
+
+    async execute(stmt: InStatement): Promise<ResultSet> {
+        const stream = this.client.openStream();
+        try {
+            const hranaStmt = stmtToHrana(stmt);
+            const hranaRows = await stream.query(hranaStmt);
+            return resultSetFromHrana(hranaRows);
+        } catch (e) {
+            throw mapHranaError(e);
+        } finally {
+            stream.close();
+        }
+    }
+
+    async batch(stmts: Array<InStatement>): Promise<Array<ResultSet>> {
+        const stream = this.client.openStream();
+        try {
+            const batch = stream.batch();
+            
+            const beginStep = batch.step();
+            const beginPromise = beginStep.run("BEGIN").catch(_ => undefined);
```

Comment:
I was under the impression that sqld was going to implicitly wrap a list of statements inside begin/commit and manage rollbacks internally if any of the statements fail.

Reply:
We decided that instead of baking the details of transactions into the protocol, sqld will provide a general mechanism to execute batches, and the clients will use it to implement non-interactive transactions.

For the users, this is just an implementation detail, the only thing that should matter for them is that the behavior is correct (the statements are executed in a transaction, which is committed only if all of the statements suceeded).